File: widgetalarm/app/src/main/java/com/daeseong/widgetalarm/AlarmWidget.java
package com.daeseong.widgetalarm;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.widget.RemoteViews;
import java.text.SimpleDateFormat;
import java.util.Date;

public class AlarmWidget extends AppWidgetProvider {

    private static final String TAG = AlarmWidget.class.getSimpleName();

    private PendingIntent pendingIntent = null;

    private long lMin = 1000 * 60 * 1;// Millisec * Second * Minute

    static void updateAppWidget(Context context, AppWidgetManager appWidgetManager, int appWidgetId) {

        Log.e(TAG, "updateAppWidget");

        try {

            RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.alarm_widget);
            remoteViews.setTextViewText(R.id.tv1, getTimeDate());
            appWidgetManager.updateAppWidget(appWidgetId, remoteViews);

        } catch (Exception ex) {
            Log.e(TAG, ex.getMessage().toString());
        }
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {

        Log.e(TAG, "onUpdate");

        //1분에 한번씩 시간 업데이트

        try {

            for (int appWidgetId : appWidgetIds) {
                updateAppWidget(context, appWidgetManager, appWidgetId);
            }

            Intent intent = new Intent(context, AlarmReceiver.class);
            pendingIntent = PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_IMMUTABLE);

            AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);

            alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, System.currentTimeMillis(), lMin, pendingIntent);

        } catch (Exception ex) {
            Log.e(TAG, ex.getMessage().toString());
        }
    }

    @Override
    public void onEnabled(Context context) {

        Log.e(TAG, "onEnabled");
    }

    @Override
    public void onDisabled(Context context) {

        Log.e(TAG, "onDisabled");

        try {

            AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
            alarmManager.cancel(pendingIntent);

        } catch (Exception ex) {
            Log.e(TAG, ex.getMessage().toString());
        }
    }

    private static String getTimeDate() {
        SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm:ss");
        return dateFormat.format(new Date());
    }
}

File: widgetalarm(kotlin)/app/src/main/java/com/daeseong/widgetalarm/AlarmWidget.kt
package com.daeseong.widgetalarm

import android.app.AlarmManager
import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.util.Log
import android.widget.RemoteViews
import java.text.SimpleDateFormat
import java.util.*

class AlarmWidget : AppWidgetProvider() {

    private val tag = AlarmWidget::class.java.simpleName

    private var pendingIntent: PendingIntent? = null

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {

        Log.e(tag, "onUpdate")

        //1분에 한번씩 시간 업데이트
        val lMin: Long = 1000 * 60 * 1 // Millisec * Second * Minute

        try {
            for (appWidgetId in appWidgetIds) {
                updateAppWidget(context, appWidgetManager, appWidgetId)
            }

            val intent = Intent(context, AlarmReceiver::class.java)
            pendingIntent = PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_IMMUTABLE)

            val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager

            alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, System.currentTimeMillis(), lMin, pendingIntent)

        } catch (ex: Exception) {
            Log.e(tag, ex.message.toString())
        }
    }

    override fun onEnabled(context: Context) {
        Log.e(tag, "onEnabled")
    }

    override fun onDisabled(context: Context) {

        Log.e(tag, "onDisabled")

        try {
            val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
            pendingIntent?.let {
                alarmManager.cancel(it)
            }
        } catch (ex: Exception) {
            Log.e(tag, ex.message.toString())
        }
    }

    fun updateAppWidget(context: Context, appWidgetManager: AppWidgetManager, appWidgetId: Int) {

        Log.e(tag, "updateAppWidget")

        try {
            val remoteViews = RemoteViews(context.packageName, R.layout.alarm_widget)
            remoteViews.setTextViewText(R.id.tv1, getTimeDate())
            appWidgetManager.updateAppWidget(appWidgetId, remoteViews)
        } catch (ex: Exception) {
            Log.e(tag, ex.message.toString())
        }
    }

    fun getTimeDate(): String {
        val dateFormat = SimpleDateFormat("HH:mm:ss")
        return dateFormat.format(Date())
    }
}
